libraries/AC_AttitudeControl/AC_PosControl.cpp

```diff
@@ -1018,36 +995,21 @@ void AC_PosControl::run_xy_controller(float dt)
     float ekfGndSpdLimit, ekfNavVelGainScaler;
     AP::ahrs_navekf().getEkfControlLimits(ekfGndSpdLimit, ekfNavVelGainScaler);
 
-    Vector3f curr_pos = _inav.get_position();
-    float kP = ekfNavVelGainScaler * _p_pos_xy.kP(); // scale gains to compensate for noisy optical flow measurement in the EKF
+    // Position Controller
 
-    // avoid divide by zero
-    if (kP <= 0.0f) {
-        _vel_target.x = 0.0f;
-        _vel_target.y = 0.0f;
-    } else {
-        // calculate distance error
-        _pos_error.x = _pos_target.x - curr_pos.x;
-        _pos_error.y = _pos_target.y - curr_pos.y;
-
-        // Constrain _pos_error and target position
-        // Constrain the maximum length of _vel_target to the maximum position correction velocity
-        // TODO: replace the leash length with a user definable maximum position correction
-        if (limit_vector_length(_pos_error.x, _pos_error.y, _leash)) {
-            _pos_target.x = curr_pos.x + _pos_error.x;
-            _pos_target.y = curr_pos.y + _pos_error.y;
-        }
-
-        _vel_target = sqrt_controller_3D(_pos_error, kP, _accel_cms);
-    }
+    Vector3f curr_pos = _inav.get_position();
+    Vector2f vel_target = _p_pos_xy.update_all(_pos_target.x, _pos_target.y, curr_pos, _leash, _accel_cms);
 
     // add velocity feed-forward
+    _vel_target.x = vel_target.x;
+    _vel_target.y = vel_target.y;
+    // acceleration to correct for velocity error and scale PID output to compensate for optical flow measurement induced EKF noise
+    _vel_target.x *= ekfNavVelGainScaler;
+    _vel_target.y *= ekfNavVelGainScaler;
```

Comment:
```suggestion
    _vel_target *= ekfNavVelGainScaler;
```
no need to do the components seperatly

Reply:
We do not apply this to the z axis as this is the XY controller component. So we can't use:
_vel_target *= ekfNavVelGainScaler;

Reply:
@lthall you're absolutely correct.

However, if @IamPete1 had suggested multiplying the local `vel_target` rather than the the object state's `_vel_target`, he'd be correct (IMO)  (i.e. multiply `vel_target` by `ekfNavVelGainScaler` before assigning it member-wise into `_vel_target`)

Reply:
100%!

Reply:
this has been done now so marking this as resolved.